sycl-jit/jit-compiler/lib/rtc/RTC.cpp

```diff
@@ -0,0 +1,168 @@
+    std::chrono::duration<double, std::milli> HashTime = Stop - Start;
+    llvm::dbgs() << "Hashing of " << SourceFile.Path << " took "
+                 << int(HashTime.count()) << " ms\n";
+  }
+
+  return RTCHashResult{Hash.c_str()};
+}
+
+JIT_EXPORT_SYMBOL RTCResult compileSYCL(InMemoryFile SourceFile,
+                                        View<InMemoryFile> IncludeFiles,
+                                        View<const char *> UserArgs,
+                                        View<char> CachedIR, bool SaveIR) {
+  llvm::LLVMContext Context;
+  std::string BuildLog;
+  configureDiagnostics(Context, BuildLog);
+
+  auto UserArgListOrErr = parseUserArgs(UserArgs);
+  if (!UserArgListOrErr) {
+    return errorTo<RTCResult>(UserArgListOrErr.takeError(),
+                              "Parsing of user arguments failed",
+                              RTCResult::RTCErrorCode::INVALID);
+  }
+  llvm::opt::InputArgList UserArgList = std::move(*UserArgListOrErr);
+
+  llvm::StringRef TraceFileName;
+  if (auto *Arg =
+          UserArgList.getLastArg(clang::driver::options::OPT_ftime_trace_EQ)) {
+    TraceFileName = Arg->getValue();
+    int Granularity =
+        500; // microseconds. Same default as in `clang::FrontendOptions`.
+    if (auto *Arg = UserArgList.getLastArg(
+            clang::driver::options::OPT_ftime_trace_granularity_EQ)) {
+      if (!llvm::to_integer(Arg->getValue(), Granularity)) {
+        BuildLog += "warning: ignoring malformed argument: '" +
+                    Arg->getAsString(UserArgList) + "'\n";
+      }
+    }
+    bool Verbose =
+        UserArgList.hasArg(clang::driver::options::OPT_ftime_trace_verbose);
+
+    llvm::timeTraceProfilerInitialize(Granularity, /*ProcName=*/"sycl-rtc",
+                                      Verbose);
+  }
+
+  std::unique_ptr<llvm::Module> Module;
+
+  if (CachedIR.size() > 0) {
+    llvm::StringRef IRStr{CachedIR.begin(), CachedIR.size()};
+    std::unique_ptr<llvm::MemoryBuffer> IRBuf =
+        llvm::MemoryBuffer::getMemBuffer(IRStr, /*BufferName=*/"",
+                                         /*RequiresNullTerminator=*/false);
+    auto ModuleOrError = llvm::parseBitcodeFile(*IRBuf, Context);
+    if (!ModuleOrError) {
+      // Not a fatal error, we'll just compile the source string normally.
+      BuildLog.append(formatError(ModuleOrError.takeError(),
+                                  "Loading of cached device code failed"));
+    } else {
+      Module = std::move(*ModuleOrError);
+    }
+  }
+
+  bool FromSource = false;
+  if (!Module) {
+    auto ModuleOrErr = compileDeviceCode(SourceFile, IncludeFiles, UserArgList,
+                                         BuildLog, Context);
+    if (!ModuleOrErr) {
+      return errorTo<RTCResult>(ModuleOrErr.takeError(),
+                                "Device compilation failed");
+    }
+
+    Module = std::move(*ModuleOrErr);
+    FromSource = true;
+  }
```

Comment:
Very much a nit, but it seems to me like you could also just do:
```suggestion
  bool FromSource = !Module;
  if (FromSource) {
    auto ModuleOrErr = compileDeviceCode(SourceFile, IncludeFiles, UserArgList,
                                         BuildLog, Context);
    if (!ModuleOrErr) {
      return errorTo<RTCResult>(ModuleOrErr.takeError(),
                                "Device compilation failed");
    }

    Module = std::move(*ModuleOrErr);
  }
```